File: go_algorithms/algorithms/RowByRowEnhanced.go
package algorithms

import (
	"fmt"
	"sync"
	"time"
)

func RowByRowEnhanced(B, C [][]int, bsize int) ([][]int, int64) {
	fmt.Println("Método: RowByRowEnhanced")

	startTime := time.Now().UnixNano()
	var wg sync.WaitGroup
	size := len(C)
	A := make([][]int, size)
	for i := range A {
		A[i] = make([]int, size)
	}
	wg.Add(2)

	go func() {
		defer wg.Done()
		for i1 := 0; i1 < size/2; i1 += bsize {
			for j1 := 0; j1 < size; j1 += bsize {
				for k1 := 0; k1 < size; k1 += bsize {
					for i := i1; i < i1+bsize && i < size; i++ {
						for j := j1; j < j1+bsize && j < size; j++ {
							for k := k1; k < k1+bsize && k < size; k++ {
								A[i][k] += B[i][j] * C[j][k]
							}
						}
					}
				}
			}
		}
	}()

	go func() {
		defer wg.Done()
		for i1 := size / 2; i1 < size; i1 += bsize {
			for j1 := 0; j1 < size; j1 += bsize {
				for k1 := 0; k1 < size; k1 += bsize {
					for i := i1; i < i1+bsize && i < size; i++ {
						for j := j1; j < j1+bsize && j < size; j++ {
							for k := k1; k < k1+bsize && k < size; k++ {
								A[i][k] += B[i][j] * C[j][k]
							}
						}
					}
				}
			}
		}
	}()

	wg.Wait()

	endTime := time.Now().UnixNano()
	executionTime := endTime - startTime
	return A, executionTime
}

File: python_algorithms/algorithms/RowByRowEnhanced.py
import numpy as np
import threading
import time
import math

def Multiply(B, C):
    print("Método: RowByRowEnhanced")

    start_time = time.time_ns()
    size = len(C)
    A = np.zeros((size, size), dtype=np.int64)

    bsize = max(1, int(math.sqrt(size)))  # Calcula bsize como la raíz cuadrada del tamaño de la matriz

    def worker(start, end):
        for i1 in range(start, end, bsize):
            for j1 in range(0, size, bsize):
                for k1 in range(0, size, bsize):
                    for i in range(i1, min(i1 + bsize, size)):
                        for j in range(j1, min(j1 + bsize, size)):
                            for k in range(k1, min(k1 + bsize, size)):
                                A[i][k] += B[i][j] * C[j][k]

    threads = []
    mid = size // 2

    thread1 = threading.Thread(target=worker, args=(0, mid))
    thread2 = threading.Thread(target=worker, args=(mid, size))

    thread1.start()
    thread2.start()

    threads.append(thread1)
    threads.append(thread2)

    for thread in threads:
        thread.join()

    end_time = time.time_ns()
    execution_time = end_time - start_time
    return A, execution_time
